app/Filament/Resources/EventResource/RelationManagers/AchievementsRelationManager.php

```diff
@@ -77,7 +82,68 @@ public function table(Table $table): Table
 
             ])
             ->actions([
-
+                Tables\Actions\ActionGroup::make([
+                    Tables\Actions\Action::make('award')
+                        ->label('Award to User(s)')
+                        ->icon('fas-trophy')
+                        ->form([
+                            Forms\Components\Textarea::make('users')
+                                ->label('CSV of User names')
+                                ->autosize()
+                                ->required(),
+                        ])
+                        ->modalHeading(function (EventAchievement $eventAchievement): string {
+                            return "Manually award {$eventAchievement->achievement->title}";
+                        })
+                        ->action(function (array $data, EventAchievement $eventAchievement): void {
+                            /** @var User $unlockedBy */
+                            $unlockedBy = auth()->user();
+
+                            $foundCount = 0;
+                            $unknown = [];
+                            $timestamp = Carbon::now();
+                            $lastFoundUser = null;
+                            foreach (explode(',', $data['users']) as $username) {
+                                $username = trim($username);
+                                if (empty($username)) {
+                                    continue;
+                                }
+
+                                // TODO: whereName
+                                $forUser = User::where('User', $username)->first();
```

Comment:
I'm sure it's already on your radar, just leaving a breadcrumb here that we should be good to cut over to `whereName` now.